docker/jupyter/custom/JupyterLockingExtension.js

```diff
@@ -0,0 +1,325 @@
+
+    //TODO URL resolution
+    const leoUrl = 'http://localhost:8080'
+        // const leoUrl = 'http://localhost:8081'
+    const welderUrl = leoUrl
+    const localizeUrl = welderUrl + '/localize'
+        // const checkMetaUrl = welderUrl + '/checkMeta'
+    const checkMetaUrl = welderUrl + '/objects/metadata'
+    const lockUrl = welderUrl + '/lock'
+    const lastLockedTimer = 60000 // in ms, should be 60000 in final PR
+
+    const headers = {
+        'Content-Type': 'application/json',
+        'Access-Control-Allow-Headers': '*'
+    }
+
+    const basePayload = {
+        // mode: "no-cors",
+        headers: headers
+    }
+
+    //TODO resolve these links
+    const jupyterBaseUrl = "http://localhost:8000/"
+    const jupyterContentsAPIUrl = jupyterBaseUrl + "api/contents/"
+
+    const getLocalPath = () => {
+        return {
+            localObjectPath: Jupyter.notebook.notebook_path
+        }
+    }
+
+    function init() {
+        checkMeta()
+        initSyncMaintainer()
+    }
+
+    async function initSyncMaintainer() {
+        var syncMaintainer = setInterval(() => {
+            checkMeta()
+        }, lastLockedTimer)
+
+        window.onbeforeunload(function() {
+            clearInterval(syncMaintainer)
+        })
+    }
+
+    async function checkMeta() {
+        const payload = Object.assign(basePayload, {
+            body: JSON.stringify(getLocalPath()),
+            method: 'POST'
+        })
+
+        return fetch(checkMetaUrl, payload)
+            .then(res => {
+                console.log('in first then of checkmeta', res)
+                if (!res.ok && res.status == 412) {
+                    promptUserWithModal(syncIssueTitle, noRemoteFileButtons, syncIssueFatalBody)
+                } else if (!res.ok && res.status >= 400) {
+                    throw Error("check metadata call failed due to status code")
+                }
+                return res.json()
+            })
+            .then(res => {
+                console.log('res in check meta after .json()', res)
+                handleCheckMetaResp(res) //sets meta state
+                toggleMetaFailureBanner(false) //sets banner for meta status
+                maintainLock(res) //gets lock 
+                return res
+            })
+            .catch(err => {
+                console.log('in checkMeta catch')
+                console.error(err)
+                removeModeBanner()
+                toggleMetaFailureBanner(true)
+            })
+    }
+
+    //this function assumes any status not included in these lists represents an in sync notebook to defend against future fields being added being auto-categorized as failures
+    function handleCheckMetaResp(res) {
+        meta = res //set meta state
+
+        const outOfSyncStatuses = ["DESYNCHRONIZED", "LOCAL_CHANGED", "REMOTE_CHANGED"]
+        const fatalSyncStatuses = ["REMOTE_NOT_FOUND"]
+
+        if (outOfSyncStatuses.includes(res.syncStatus)) {
```

Comment:
which of these 2 cases should have priority?

Reply:
Not sure I understand, they're mutually exclusive

Reply:
ha.  that is true.  never mind.